system/mp/mpcomm_mpi.cpp

```diff
@@ -0,0 +1,302 @@
+                if (hpcc_mpi::getCommStatus(req) == hpcc_mpi::CommStatus::ERROR)
+                    return false;
+                hpcc_mpi::releaseComm(req);
+            }
+        }
+       
+        return true;
+    }
+
+    bool recv(CMessageBuffer &mbuf, rank_t srcrank, mptag_t tag, rank_t *sender, unsigned timeout=MP_WAIT_FOREVER){
+        CTimeMon tm(timeout);
+        hpcc_mpi::CommRequest req = hpcc_mpi::readData(srcrank, tag, mbuf, *group, true);
+        unsigned remaining;
+        hpcc_mpi::CommStatus stat;
+        do {
+            usleep(100);
+            stat = hpcc_mpi::getCommStatus(req);
+        } while (stat == hpcc_mpi::CommStatus::INCOMPLETE && !tm.timedout(&remaining));
+//        printf("recv while existed %d\n",stat);
+        hpcc_mpi::releaseComm(req);
+//        printf("req released %d\n",stat);
+        return (stat == hpcc_mpi::CommStatus::SUCCESS);
+    }
+    
+    void barrier(void){
+        hpcc_mpi::barrier(*group);
+    }
+
+    unsigned probe(rank_t srcrank, mptag_t tag, rank_t *sender, unsigned timeout=0){
+        if (hpcc_mpi::hasIncomingMessage(srcrank, tag, *group)){
+            if (sender)
+                *sender = srcrank;
+        }
+        return false;
+    }
+    
+    void flush(mptag_t tag){
+        // Handled by MPI
+    }
+
+    bool sendRecv(CMessageBuffer &mbuff, rank_t sendrank, mptag_t sendtag, unsigned timeout=MP_WAIT_FOREVER){
+        //TODO share timeout between send/recv?
+        mptag_t replytag = createReplyTag();
+        CTimeMon tm(timeout);
+        mbuff.setReplyTag(replytag);
+        unsigned remaining;
+        if (tm.timedout(&remaining))
+            return false;
+        if (!send(mbuff,sendrank,sendtag,remaining)||tm.timedout(&remaining))
+            return false;
+        mbuff.clear();
+        return recv(mbuff,sendrank,replytag,NULL,remaining);
+    }
+
+    bool reply(CMessageBuffer &mbuf, unsigned timeout=MP_WAIT_FOREVER){
+        mptag_t replytag = mbuf.getReplyTag();
+        rank_t dstrank = group->rank(mbuf.getSender());
+        if (dstrank!=RANK_NULL) {
+            if (send (mbuf, dstrank, replytag,timeout)) {
+                mbuf.setReplyTag(TAG_NULL);
+                return true;
+            }
+            return false;
+        }
+        return false;
+    }
+
+    void cancel(rank_t srcrank, mptag_t tag){
+        assertex(srcrank!=RANK_NULL);       
+        hpcc_mpi::CommRequest req = group->getCommRequest(srcrank, tag);
+        if (req >= 0){
+            hpcc_mpi::cancelComm(req);
+        }
+    }
+
+    bool verifyConnection(rank_t rank,  unsigned timeout){
+        UNIMPLEMENTED;
+    }
+
+    bool verifyAll(bool duplex, unsigned timeout){
+        UNIMPLEMENTED;
+    }
+    
+    void disconnect(INode *node){
+        UNIMPLEMENTED;
+    }
+
+    IGroup &queryGroup() { 
+        UNIMPLEMENTED;
+    }
+    
+    IGroup *getGroup()  { 
+        return group;
```

Comment:
getGroup/queryGroup both return same thing, it's just that getXXX implies it links, so this should too and queryGroup should return without linking.

Reply:
Got it. will update.